external/mlir-hal/lib/Dialect/MHAL/Transforms/Prefill.cpp

```diff
@@ -0,0 +1,140 @@
+                      APFloat::rmNearestTiesToEven, &lostInfo);
+
+      auto constValue = builder.getFloatAttr(elementType, apValue);
+      constantOp =
+          builder.create<arith::ConstantOp>(loc, elementType, constValue);
+    } else {
+      APInt apValue(elementType.getIntOrFloatBitWidth(), initValue, true);
+      auto constValue = builder.getIntegerAttr(elementType, apValue);
+      constantOp =
+          builder.create<arith::ConstantOp>(loc, elementType, constValue);
+    }
+    return constantOp;
+  }
+
+  void insertPrefillOps(OpBuilder &builder, gpu::LaunchFuncOp &launchOp) {
+    auto func = llvm::cast<func::FuncOp>(launchOp->getParentOp());
+    auto module = llvm::cast<ModuleOp>(func->getParentOp());
+
+    auto kernel = launchOp.getKernel();
+    auto *callee = module.lookupSymbol(kernel);
+    assert(callee != nullptr && "expect to find the function defenition");
+
+    auto llvmFunc = llvm::cast<LLVM::LLVMFuncOp>(callee);
+    auto gpuModule = llvm::cast<gpu::GPUModuleOp>(llvmFunc->getParentOp());
+    assert(gpuModule != nullptr &&
+           "expect LLVMFuncOp to be inside GPUModuleOp");
+
+    SmallVector<mhal::PrefillAttr, 4> prefillAttrs;
+    if (auto moduleAttr = gpuModule->getAttr(llvmFunc.getSymName())) {
+      if (auto arrayAttr = llvm::dyn_cast<ArrayAttr>(moduleAttr)) {
+        for (auto attr : arrayAttr) {
+          if (auto prefillAttr = llvm::dyn_cast<mhal::PrefillAttr>(attr)) {
+            prefillAttrs.push_back(prefillAttr);
+          }
+        }
+      }
+    }
+
+    auto loc = builder.getUnknownLoc();
+    auto kernelOperands = launchOp.getKernelOperands();
+    for (auto attr : prefillAttrs) {
+      auto argIdx = attr.getArgIndex();
+      assert(argIdx < kernelOperands.size() &&
+             "provided arg index is out of bounds");
+
+      auto arg = kernelOperands[argIdx];
+      auto type = arg.getType().cast<MemRefType>();
+      auto elementType = type.getElementType();
+
+      builder.setInsertionPointToStart(arg.getParentBlock());
+      auto initConstant =
+          createConstantOp(builder, elementType, attr.getInitValue());
+
+      builder.setInsertionPoint(launchOp->getBlock(),
+                                --Block::iterator(launchOp));
+      builder.create<gpu::MemsetOp>(loc, mlir::Type{}, mlir::ValueRange{}, arg,
+                                    initConstant);
+    }
+  }
+
+public:
+  // Inspects each gpu:: LaunchFuncOp, find arguments which must be prefilled
+  // beforehand, and inserts the corresponding gpu::MemsetOp
+  void runOnOperation() override {
+    func::FuncOp func = getOperation();
+
+    llvm::SmallVector<gpu::LaunchFuncOp, 4> ops;
+    func.walk([&ops](gpu::LaunchFuncOp launchOp) { ops.push_back(launchOp); });
+
+    {
+      OpBuilder builder(&getContext());
+      OpBuilder::InsertionGuard insertGuard(builder);
+      for (gpu::LaunchFuncOp launchOp : ops) {
+        insertPrefillOps(builder, launchOp);
+      }
+    }
+  }
+};
+} // namespace
```

Comment:
Style nit: move method implementations outside of the `namespace{}`

Reply:
done